apps/src/music/blockly/extensions.js

```diff
@@ -185,3 +188,28 @@ export const fieldSoundsValidator = function () {
     return newValue;
   });
 };
+
+/**
+ * Extension to blocks with pattern fields that validates new values.
+ */
+export const fieldPatternsValidator = function () {
+  // A block may have a pattern field or pattern AI field, but should not have both.
+  const patternField =
+    this.getField(FIELD_PATTERN_NAME) || this.getField(FIELD_PATTERN_AI_NAME);
+
+  /**
+   * Removes invalid event notes from pattern field values.
+   * @param newValue The new instrument event value
+   * @returns The modified instrument event value
+   */
+  patternField?.setValidator(newValue => {
+    newValue.events = newValue.events.filter(
+      event =>
+        // Remove events with notes that are outside the range of our supported drum sounds.
+        event.note < PATTERN_NUM_NOTES &&
+        // Remove event with ticks that are outside the expected tick range.
+        event.tick <= newValue.length * PATTERN_NUM_EVENTS
```

Comment:
Should this be `PATTERN_AI_NUM_EVENTS` when it's a `PLAY_PATTERN_AI_AT_CURRENT_LOCATION_SIMPLE2` block?

Reply:
It actually doesn't need to be, because for an AI block `.length` will be 2.

Reply:
Ah, in that case I think we could do something a bit more special when we declare `PATTERN_NUM_EVENTS`.  The way it's used here, it's not just a shorter variant of `PATTERN_AI_NUM_EVENTS`, but in fact means something like `PATTERN_NUM_EVENTS_PER_MEASURE`.  

In fact, maybe we can declare that, and build the other ones off of it, like this?
```
export const PATTERN_NUM_EVENTS_PER_MEASURE = 16;

export const PATTERN_NUM_EVENTS = PATTERN_NUM_EVENTS_PER_MEASURE * 1;
export const PATTERN_AI_NUM_EVENTS = PATTERN_NUM_EVENTS_PER_MEASURE * 2;
```

Reply:
+1, maybe `TICKS_PER_MEASURE` since we use the same paradigm in chord/tune blocks too (i.e. `InstrumentTickEvent` referring to a single 16th note). 

(_Technicallyyyy_ tick refers to a subdivision smaller than 16th notes used in MIDI sequencing/DAWs, but since we've been using it to mean 16ths in our codebase hopefully the meaning is clear).

Reply:
Updated!